## Exploring Hacker News Posts
### 1. Introduction
Hacker News is a website started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hundreds of thousands of vistiors may view the top rated of these posts as Hacker News is extremely popular in technology and startup circles.

The original dataset can be found [here](https://www.kaggle.com/hacker-news/hacker-news-posts), do note that it has been reduced from almost 300,000 rows to approximately 20,00 rows by removing all submissions that did not recieve any comments and then randomly samplying from the remaining submissions.

Below are descriptions of the columns:

   - id: The unique identifier from Hacker News for the post
   - title: The title of the post
   - url: The URL that the posts links to, if it the post has a URL
   - num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
   - num_comments: The number of comments that were made on the post
   - author: The username of the person who submitted the post
   - created_at: The date and time at which the post was submitted

In [1]:
#Opening the dataset
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)

In [2]:
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


### 2. Removing the Header from the data and exploring the data

In [3]:
headers = hn[:1] #Copy Header information to it's own variable
hn = hn[1:] #set the data to just be the data without the header

In [4]:
headers

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]

In [5]:
hn[:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

### 3. Extracting Ask / Show / Other posts from the Hacker News posts.


In [6]:
ask_posts = []
show_posts = []
other_posts = []

In [7]:
for row in hn:
    title = row[1]
    title= title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [8]:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


Upon extracting the ask/show/other posts we can see the following... <br>
Ask   - 1744 posts<br>
Show  - 1162 posts<br>
Other - 17194 posts

### 4. Calculating the Average Number of Comments for Ask / Show HN posts.

In [9]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments+=num_comments
avg_ask_comments = total_ask_comments/len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [10]:
total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments+=num_comments
avg_show_comments = total_show_comments/len(show_posts)
print(avg_show_comments)

10.31669535283993


Upon inspecting the average comments for ask / show hn posts we can see that on average ask posts recieve roughly 4 more comments per post at 14.038 vs the average of show comment posts being 10.32. Since ask posts are more likely to recieve comments we will focus our remaining analysis on just these posts.

### 5. Find the Amount of Ask Posts and Comments by Hour Created

In [11]:
import datetime as dt
result_list=[]

for row in ask_posts:
    result_list.append([row[6],int(row[4])])
counts_by_hour={}
comments_by_hour={}
for row in result_list:
    date = dt.datetime.strptime(row[0],"%m/%d/%Y %H:%M")
    hour = dt.datetime.strftime(date,"%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = row[1]
    else:
        counts_by_hour[hour]+=1
        comments_by_hour[hour]+=row[1]
        
    

In [12]:
print(comments_by_hour)
print(counts_by_hour)

{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}
{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


Exploring the data we found we can see that certain hours are more popular by hour. Hour 15 with 4477 seems to be an incredible popular time to comment. While hour 04 is much less at 337.

### 6. Calculating the Average Number of Comments for Ask HN Posts by Hour

In [13]:
avg_by_hour = []
for hours in comments_by_hour:
    avg_by_hour.append([hours,comments_by_hour[hours]/counts_by_hour[hours]])

In [14]:
avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

We can see after calculating the avg number of comments per hour that roughly every post by hour averages about 7 comments roughly. So next we will clean up and sort this data to make it alittle easier to read.

### 7. Sorting and Printing values from a list of lists

In [15]:
## Invert the Comments with the hour
swap_avg_by_hour = []
for element in avg_by_hour:
    swap_avg_by_hour.append([element[1],element[0]])
swap_avg_by_hour

[[5.5777777777777775, '09'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [16.796296296296298, '16'],
 [7.985294117647059, '23'],
 [9.41095890410959, '12'],
 [11.46, '17'],
 [38.5948275862069, '15'],
 [16.009174311926607, '21'],
 [21.525, '20'],
 [23.810344827586206, '02'],
 [13.20183486238532, '18'],
 [7.796296296296297, '03'],
 [10.08695652173913, '05'],
 [10.8, '19'],
 [11.383333333333333, '01'],
 [6.746478873239437, '22'],
 [10.25, '08'],
 [7.170212765957447, '04'],
 [8.127272727272727, '00'],
 [9.022727272727273, '06'],
 [7.852941176470588, '07'],
 [11.051724137931034, '11']]

In [16]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [17]:
print("Top 5 Hours for Ask Posts Comments")
i = 0
for row in sorted_swap:
    if i < 5:
        date = dt.time(hour=int(row[1]))
        hour = dt.time.strftime(date,"%H:%M")
        print("{hour}:{avg:.2f} avg comments per post".format(hour=hour, avg=row[0]))
        i+=1

Top 5 Hours for Ask Posts Comments
15:00:38.59 avg comments per post
02:00:23.81 avg comments per post
20:00:21.52 avg comments per post
16:00:16.80 avg comments per post
21:00:16.01 avg comments per post


3 pm = 38.6<br>
2 am = 23.8<br>
8 pm = 21.5<br>
4 pm = 16.8<br>
9 pm = 16<br>

Seemingly the best hours to get a higher chance of recieving comments take place in the afternoon to night for EST. There seems to be a correlation with just before the work day typically ends for most that leads to a higher amount of comments.

### 8. Calculating the Average Number of Points for Ask / Show HN posts.

In [18]:
total_ask_points = 0
for row in ask_posts:
    num_points = int(row[3])
    total_ask_points+=num_points
avg_ask_points = total_ask_points/len(ask_posts)
print(avg_ask_points)

15.061926605504587


In [19]:
total_show_points = 0
for row in ask_posts:
    num_points = int(row[3])
    total_show_points+=num_points
avg_show_points = total_show_points/len(show_posts)
print(avg_show_points)

22.605851979345957


Upon inspecting the average points for ask / show hn posts we can see that on average show posts recieve roughly 7 more points per post at 22.61 vs the average of ask points per post being 10.32. Since show posts are more likely to recieve points we will focus our remaining analysis on just these posts.

### 9. Finding the Amount of Show Posts and Points by Hour Created

In [20]:
import datetime as dt
result_list=[]

for row in ask_posts:
    result_list.append([row[6],int(row[3])])
count_by_hour={}
points_by_hour={}
for row in result_list:
    date = dt.datetime.strptime(row[0],"%m/%d/%Y %H:%M")
    hour = dt.datetime.strftime(date,"%H")
    if hour not in count_by_hour:
        count_by_hour[hour] = 1
        points_by_hour[hour] = row[1]
    else:
        count_by_hour[hour]+=1
        points_by_hour[hour]+=row[1]

In [21]:
print(points_by_hour)
print(count_by_hour)

{'09': 329, '13': 2062, '10': 1102, '14': 1282, '16': 2522, '23': 581, '12': 782, '17': 1941, '15': 3479, '21': 1721, '20': 1151, '02': 793, '18': 1741, '03': 374, '05': 552, '19': 1513, '01': 700, '22': 511, '08': 515, '04': 389, '00': 451, '06': 591, '07': 361, '11': 825}
{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


Exploring the data we found we can see that certain hours are more popular by hour. Of course being that hour 15 was the most popular hour in our original exploration of comments per hour no surprise to see that hour is also one of the most popular for points as well.

### 10.Calculating the Average Number of Points for Show HN Posts by Hour

In [22]:
avg_by_hour = []
for hours in points_by_hour:
    avg_by_hour.append([hours,points_by_hour[hours]/count_by_hour[hours]])
avg_by_hour

[['09', 7.311111111111111],
 ['13', 24.258823529411764],
 ['10', 18.677966101694917],
 ['14', 11.981308411214954],
 ['16', 23.35185185185185],
 ['23', 8.544117647058824],
 ['12', 10.712328767123287],
 ['17', 19.41],
 ['15', 29.99137931034483],
 ['21', 15.788990825688073],
 ['20', 14.3875],
 ['02', 13.672413793103448],
 ['18', 15.972477064220184],
 ['03', 6.925925925925926],
 ['05', 12.0],
 ['19', 13.754545454545454],
 ['01', 11.666666666666666],
 ['22', 7.197183098591549],
 ['08', 10.729166666666666],
 ['04', 8.27659574468085],
 ['00', 8.2],
 ['06', 13.431818181818182],
 ['07', 10.617647058823529],
 ['11', 14.224137931034482]]

We can see after calculating the avg number of points per hour that roughly every post by hour averages atleast about 7 comments roughly as well. So very similar to the Ask comments total. So next we will clean up and sort this data to make it alittle easier to read.

### 11. Sorting and Printing values from a list of lists

In [23]:
## Invert the Points with the hour
swap_avg_by_hour = []
for element in avg_by_hour:
    swap_avg_by_hour.append([element[1],element[0]])
swap_avg_by_hour

[[7.311111111111111, '09'],
 [24.258823529411764, '13'],
 [18.677966101694917, '10'],
 [11.981308411214954, '14'],
 [23.35185185185185, '16'],
 [8.544117647058824, '23'],
 [10.712328767123287, '12'],
 [19.41, '17'],
 [29.99137931034483, '15'],
 [15.788990825688073, '21'],
 [14.3875, '20'],
 [13.672413793103448, '02'],
 [15.972477064220184, '18'],
 [6.925925925925926, '03'],
 [12.0, '05'],
 [13.754545454545454, '19'],
 [11.666666666666666, '01'],
 [7.197183098591549, '22'],
 [10.729166666666666, '08'],
 [8.27659574468085, '04'],
 [8.2, '00'],
 [13.431818181818182, '06'],
 [10.617647058823529, '07'],
 [14.224137931034482, '11']]

In [24]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[29.99137931034483, '15'],
 [24.258823529411764, '13'],
 [23.35185185185185, '16'],
 [19.41, '17'],
 [18.677966101694917, '10'],
 [15.972477064220184, '18'],
 [15.788990825688073, '21'],
 [14.3875, '20'],
 [14.224137931034482, '11'],
 [13.754545454545454, '19'],
 [13.672413793103448, '02'],
 [13.431818181818182, '06'],
 [12.0, '05'],
 [11.981308411214954, '14'],
 [11.666666666666666, '01'],
 [10.729166666666666, '08'],
 [10.712328767123287, '12'],
 [10.617647058823529, '07'],
 [8.544117647058824, '23'],
 [8.27659574468085, '04'],
 [8.2, '00'],
 [7.311111111111111, '09'],
 [7.197183098591549, '22'],
 [6.925925925925926, '03']]

In [25]:
print("Top 5 Hours for Show Posts Points")
i = 0
for row in sorted_swap:
    if i < 5:
        date = dt.time(hour=int(row[1]))
        hour = dt.time.strftime(date,"%H:%M")
        print("{hour}:{avg:.2f} avg points per post".format(hour=hour, avg=row[0]))
        i+=1

Top 5 Hours for Show Posts Points
15:00:29.99 avg points per post
13:00:24.26 avg points per post
16:00:23.35 avg points per post
17:00:19.41 avg points per post
10:00:18.68 avg points per post


3 pm = 29.99<br>
1 pm = 24.26<br>
4 pm = 23.35<br>
5 pm = 19.41<br>
10 am = 18.68<br>

Seemingly the best hours to get a higher chance of recieving points take place in the  late moring to early afternoon for EST. There seems to be a correlation with lunch time to the winding down of the work day that leads to a higher amount of points.